update python devel docs (#77499)
ansible/ansible: docs/docsite/rst/dev_guide/developing_python_3.rst

Ansible and Python 3
********************

> The ``ansible-core`` code runs Python 3 (for specific versions check ( ../../installation_guide/intro_installation.html#control-node-requirements). Contributors to ansible-core and to Ansible Collections should be aware of the tips in this document so that they can write code that will run on the same versions of Python as the rest of Ansible.

Reviewer: ```suggestion
The ``ansible-core`` code runs Python 3 (for specific versions check :file:`../../installation_guide/intro_installation.html#control-node-requirements`). Contributors to ``ansible-core`` and to Ansible Collections should be aware of the tips in this document so that they can write code that will run on the same versions of Python as the rest of Ansible.
```

.. contents::
   :local:

We do have some considerations depending on the types of Ansible code:

> 1. controller-side code - code that runs on the machine where you invoke :command:`/usr/bin/ansible`, only needs to support the controller's Python versions

Reviewer: ```suggestion
1. controller-side code - code that runs on the machine where you invoke :command:`/usr/bin/ansible`, only needs to support the controller's Python versions.
```

> 2. modules - the code which Ansible transmits to and invokes on the managed machine. They need to support the 'managed node' Python versions, with some exceptions.

Reviewer: ```suggestion
2. modules - the code which Ansible transmits to and invokes on the managed machine. Modules need to support the 'managed node' Python versions, with some exceptions.
```

> 3. shared ``module_utils`` code - the common code that's used by modules to perform tasks and sometimes used by controller-side code as well, so they need to support the same range of Python as the modules.

Reviewer: ```suggestion
3. shared ``module_utils`` code - the common code that is  used by modules to perform tasks and sometimes used by controller-side code as well. Shared ``module_utils`` code needs to support the same range of Python as the modules.
```

However, the three types of code do not use the same string strategy. If you're developing a module or some ``module_utils`` code, be sure to read the section on string strategy carefully.

.. note:

>     - While modules can be written in any language, the above applies to code contributed to the core project, which only supports specific Python versions (and Powershell for Windows)

Reviewer: ```suggestion
    - While modules can be written in any language, the above applies to code contributed to the core project, which only supports specific Python versions and Powershell for Windows.
```

Minimum version of Python 3.x and Python 2.x
============================================

> See  ../../installation_guide/intro_installation.html#control-node-requirements and  ../../installation_guide/intro_installation.html#control-node-requirements for the

Reviewer: Looks like this is pointing to the same thing twice?  The correct linking would be :ref:`control_node_requirements` but not sure if you wanted to point to someplace else as well?

specific versions supported.

> Your custom modules can support any version of Python (or other languages) you want, but the above are the requirements for the code contributed to the Ansible project.

Reviewer: to the Ansible project to to ansible-core?

Author: afaik, both?

Controller string strategy: the Unicode Sandwich
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Until recently the core supported Python 2.x and followed this strategy, known as the Unicode Sandwich (named

Reviewer: ```suggestion
Until recently ``ansible-core`` supported Python 2.x and followed this strategy, known as the Unicode Sandwich (named
```
